### PR TITLE
feat(edit-ema): Add Loading State #27305

### LINKED ISSUE #27305: Edit EMA: Add Loading Indicators for Improved Edit Mode UX
### User Story
As a dotCMS user, I want to see loading indicators in the edit mode when certain actions are happening, such as loading a remote page, changing query parameters (language or persona), adding a new piece of content, or deleting a piece of content. This will provide me with visual feedback and improve my user experience by letting me know that actions are being processed in the background.

Needs to add loading indicator when:

```[tasklist]
### Acceptance Criteria
- [x] Edit mode is loading a remote page
- [x] Language change
- [x] Persona change
- [x] Add a piece of content from add button
- [x] Add a piece of content from palette
- [x] Remove a piece of content
```

### Proposed Objective
Static content: "Core Features"

### Proposed Priority
Static content: "Priority 3 - Average"

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.:
Maybe we can use this one for the page https://www.primefaces.org/primeng-v15-lts/progressbar#indeterminate

### Assumptions & Initiation Needs
- The dotCMS platform has the capability to display loading indicators.
- The design and placement

### PR DESCRIPTION
**Pull request by zJaaal**
## Proposed Changes
* Add editorState to store to handle the loading state
* Add mutationObserver to handle add/remove of the nodes in NextJS
* Add new event to handle the loaded state when add/remove nodes in nextJS

## Screenshot

https://github.com/dotCMS/core/assets/63567962/67062149-a66b-4ed9-a5c3-bdbe1f6a4619

### CODE REVIEW

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html [hunk 1]
```diff
@@ -93,7 +106,7 @@
                 (addContent)="addContentlet($event)"
                 data-testId="contentlet-tools" />
             <dot-ema-page-dropzone
-                *ngIf="!!rows.length && !!dragItem && !currentDevice"
```
Comment (fmontes) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html:96]: I think that this `dragItem` needs to be here but we are loosing it from a bad merge, I think is related with this https://github.com/dotCMS/core/pull/27290

File: core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.ts [hunk 2]
```diff
@@ -30,6 +31,7 @@ export interface EditEmaState {
     error?: number;
     editor: DotPageApiResponse;
     isEnterpriseLicense: boolean;
+    editorState: EDITOR_STATE;
```
[the remark this answers is not in the document]
Comment (zJaaal) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.ts:34]: Added to todo list

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html (hunk 1 above)
Comment (zJaaal) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html:109]: added

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts [hunk 3]
```diff
@@ -539,9 +545,19 @@ export class EditEmaEditorComponent implements OnInit, OnDestroy {
...
+
             [CUSTOMER_ACTIONS.SET_URL]: () => {
                 const payload = <SetUrlPayload>data.payload;
 
+                if (this.activatedRouter.snapshot.queryParams.url === payload.url) {
+                    this.store.updateEditorState(EDITOR_STATE.LOADED);
+                } else {
+                    this.store.updateEditorState(EDITOR_STATE.LOADING);
+                }
+
```
Comment (fmontes) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:560]: Add comments here explaining why of this... we're going to forget about this in 2 days 🤣

File: ema/nextjs/src/components/layout/navigation.js [hunk 4]
```diff
@@ -1,15 +1,24 @@
...
+
     return (
         <nav className={className}>
             <ul className="flex space-x-4 text-white">
                 <li>
-                    <Link href="/">Home</Link>
+                    <Link href={{
+                        pathname: '/',
+                        query: Object.fromEntries(searchParams.entries())
+                    }}>Home</Link>
```
Comment (fmontes) [inline ema/nextjs/src/components/layout/navigation.js:14]: Add comments here why were doing this as well.

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts (hunk 3 above)
Comment (zJaaal) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:560]: Done

File: ema/nextjs/src/components/layout/navigation.js (hunk 4 above)
Comment (zJaaal) [reply ema/nextjs/src/components/layout/navigation.js:14]: done

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.ts [hunk 5]
```diff
@@ -386,7 +409,13 @@ export class EditEmaStore extends ComponentStore<EditEmaState> {
...
 
     private createPageURL(params: DotPageApiParams): string {
-        const url = params.url.startsWith('/') ? params.url.slice(1) : params.url;
+        const url = params.url
+            .replace(/^\//g, '')
+            .split('/')
+            .filter((part, i) => {
+                return !i || part !== 'index'; // Filter the index from the url if it is at the last position
+            })
+            .join('/');
```
[the remark this answers is not in the document]
Comment (zJaaal) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.ts]: now we are

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts [hunk 6]
```diff
@@ -539,9 +545,25 @@ export class EditEmaEditorComponent implements OnInit, OnDestroy {
...
+                // When we set the url, we trigger in the shell component a load to get the new state of the page
+                // This triggers a rerender that makes nextjs to send the set_url again
+                // But this time the params are the same so the shell component wont trigger a load and there we know that the page is loaded
+
+                if (this.activatedRouter.snapshot.queryParams.url === payload.url) {
+                    this.store.updateEditorState(EDITOR_STATE.LOADED);
+                } else {
+                    this.store.updateEditorState(EDITOR_STATE.LOADING);
+                }
+
```
[the remark this answers is not in the document]
Comment (zJaaal) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:566]: Done!

File: core-web/libs/sdk/react/src/lib/hooks/usePageEditor.tsx [hunk 7]
```diff
@@ -69,9 +69,37 @@ export const usePageEditor = (
...
+            }
+        });
+
+        observer.current?.observe(document, { childList: true, subtree: true });
+
+        return () => {
+            if (observer.current) observer.current.disconnect();
+        };
+    }, []);
+
```
[the remark this answers is not in the document]
Comment (zJaaal) [reply core-web/libs/sdk/react/src/lib/hooks/usePageEditor.tsx:102]: Not worth the effort for now.

**Review by fmontes — APPROVED** [review]

**Review by manuelrojas — APPROVED** [review]
LGTM

**Pull request merged**